rsocket/RSocketServiceHandler.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+#include "rsocket/RSocketException.h"
+#include "rsocket/RSocketNetworkStats.h"
+#include "rsocket/RSocketParameters.h"
+#include "rsocket/RSocketResponder.h"
+#include "rsocket/RSocketServerState.h"
+#include "rsocket/RSocketStats.h"
+#include "rsocket/internal/Common.h"
+
+namespace rsocket {
+
+// This is a convenience function which is used to create a simplified
+// RSocketServiceHandler.
+using OnNewSetupFn =
+    folly::Function<std::shared_ptr<RSocketResponder>(const SetupParameters&)>;
+
+// This struct holds all the necessary information needed by the RSocketServer
+// to initiate a connection with a client.
+struct RSocketConnectionParams {
+  RSocketConnectionParams(
+      std::shared_ptr<RSocketResponder> _responder,
+      std::shared_ptr<RSocketStats> _stats = RSocketStats::noop(),
+      std::shared_ptr<RSocketNetworkStats> _networkStats = nullptr)
+      : responder(std::move(_responder)),
+        stats(std::move(_stats)),
+        networkStats(std::move(_networkStats)) {}
+  std::shared_ptr<RSocketResponder> responder;
+  std::shared_ptr<RSocketStats> stats;
+  std::shared_ptr<RSocketNetworkStats> networkStats;
+};
+
+// This class has to be implemented by the application.  The methods can be
+// called from different threads and it is the application's responsibility to
+// ensure thread-safety.
+class RSocketServiceHandler {
+ public:
+  virtual ~RSocketServiceHandler() = default;
+
+  // This method gets called for each client that connects to the server.  The
+  // application has to implement this method.  If the application does not
+  // want to accept this connection, it should return a RSocketException.  If
+  // the application returns a RSocketException, then an ERROR frame with code
+  // REJECTED_SETUP is sent to the client.  The exception message is sent as
+  // payload.
+  virtual folly::Expected<RSocketConnectionParams, RSocketException> onNewSetup(
```

Comment:
What you are proposing is similar to what we had previously (the application called `RSocketSetup.createRSocketStateMachine()`).  We had extensive discussion on this and decided to go with this approach.  The drawback with the old approach is that the application has to remember to call the right method and it will be repeated across all applications.  Making the application call a protocol-method also seemed like back-and-forth.  The old approach is optimized more for the case where the application wants to cold-resume.

But we expect most applications using RSocket to be simple (without leveraging cold-resumption).  In which case they have to worry only about the `onNewSetup()` method.  The current design is optimized to make the most common use-case simpler.